### ISSUE TITLE
Model is setting sub document array as empty objects

### ISSUE THREAD

**Issue by abubakkarbutt** [Automattic/mongoose#6141]
```
import mongoose from "mongoose";
const PetSchema = mongoose.Schema({name: String});
const PersonSchema = mongoose.Schema({name: String, pets: [PetSchema]});

const PersonModel = mongoose.model("Person", PersonSchema);

const john = new PersonModel({
  name: "John Doe",
  pets: [
    {name: "Parrot"},
    {name: "Mouse"},
    {name: "Dog"},
  ]
});

console.log(john);
```
Result:
```
    { name: 'John Doe',
      pets: [ {}, {}, {} ],
      _id: 5a8639ec5103f9066c659d9e }
```
It works when I import mongoose with require()
`const mongoose = require("mongoose");`
I don't know what is causing different results

**Comment by abubakkarbutt** [comment]
I have to rollback to version 5.0.4 to fix the issue.

**Comment by lineus** [comment]
I tried this and got the following results:
```
InspiredMacPro:6141 lineus$ cat 6141.js
import mongoose from 'mongoose'

console.log('version: ', mongoose.version)
const PetSchema = mongoose.Schema({ name: String })
const PersonSchema = mongoose.Schema({ name: String, pets: [PetSchema] })

const PersonModel = mongoose.model('Person', PersonSchema)

const john = new PersonModel({
  name: 'John Doe',
  pets: [
    { name: 'Parrot' },
    { name: 'Mouse' },
    { name: 'Dog' }
  ]
})

console.log(john)
```
```
InspiredMacPro:6141 lineus$ npx babel 6141.js -o index.js --presets env
InspiredMacPro:6141 lineus$ node index.js
version:  5.0.4
{ pets:
   [ { _id: 5a8731316d82aa631e43919c, name: 'Parrot' },
     { _id: 5a8731316d82aa631e43919b, name: 'Mouse' },
     { _id: 5a8731316d82aa631e43919a, name: 'Dog' } ],
  _id: 5a8731316d82aa631e439199,
  name: 'John Doe' }
InspiredMacPro:6141 lineus$ mv index.js 5.0.4.index.js
```
```
InspiredMacPro:6141 lineus$ npm install mongoose@5.0.5

InspiredMacPro:6141 lineus$ npx babel 6141.js -o index.js --presets env
InspiredMacPro:6141 lineus$ node index.js
version:  5.0.5
{ name: 'John Doe',
  pets:
   [ { name: 'Parrot', _id: 5a8731635007806360e2dc72 },
     { name: 'Mouse', _id: 5a8731635007806360e2dc71 },
     { name: 'Dog', _id: 5a8731635007806360e2dc70 } ],
  _id: 5a8731635007806360e2dc73 }
InspiredMacPro:6141 lineus$ mv index.js 5.0.5.index.js
```
```
InspiredMacPro:6141 lineus$ npx babel 6141.js -o index.js --presets env
InspiredMacPro:6141 lineus$ node index.js
version:  5.0.6
{ name: 'John Doe',
  pets:
   [ { name: 'Parrot', _id: 5a873198b1d50c63a531acac },
     { name: 'Mouse', _id: 5a873198b1d50c63a531acab },
     { name: 'Dog', _id: 5a873198b1d50c63a531acaa } ],
  _id: 5a873198b1d50c63a531acad }
InspiredMacPro:6141 lineus$ mv index.js 5.0.6.index.js
InspiredMacPro:6141 lineus$ diff 5.0.4.index.js 5.0.5.index.js
InspiredMacPro:6141 lineus$ diff 5.0.5.index.js 5.0.6.index.js
```

**Comment by abubakkarbutt** [comment]
But then why it worked by just going back to version 5.0.4. I didn't have to change any code.

**Comment by abubakkarbutt** [comment]
By the way I am using OS windows 10.
Node version. 9.xx

**Comment by lineus** [comment]
@abubakkarbutt which transpiler (including version) are you using? I just chose babel arbitrarily.
```
InspiredMacPro:5 lineus$ babel --version
6.26.0 (babel-core 6.26.0)
```

**Comment by abubakkarbutt** [comment]
I am using 
babel-core 6.23.0
babel-loader 7.1.2
webpack-preset.react-app 3.1.1

**Comment by abubakkarbutt** [comment]
I am using 
babel-core 6.26.0
babel-loader 7.1.2
webpack-preset.react-app 3.1.1

**Comment by lineus** [comment]
i found [babel-preset.react-app](https://www.npmjs.com/package/babel-preset-react-app) on npm but not webpack-preset.react-app. I tried transpiling with that preset and it didn't work at all:
```
InspiredMacPro:6141 lineus$ npx babel 6141.js -o 5.0.4.index.js --presets react-app
InspiredMacPro:6141 lineus$ head -1 5.0.4.index.js
import mongoose from 'mongoose';
InspiredMacPro:6141 lineus$
```

are you getting any errors or warnings when you build? 

Can you create a test folder and setup npm there to mirror your current build setup and then try to build just the one file (6141.js) from my first comment using mongoose > 5.04 ? If you get the same results as your initial report, create a [gist](https://gist.github.com) with the following:
  1. package.json
  1. relevant build configs ( .babelrc, webpack.config.js )
  1. the command you're using to build as well as the output

I'll do my best to recreate your build environment locally and see if I can help figure out what's going on.

**Comment by vkarpov15** [comment]
I agree with @lineus , in order to reproduce this issue we're going to need your build configs and the commands you use to build, otherwise we're just taking shots in the dark.

**Comment by abubakkarbutt** [comment]
Configurations are pretty much same as in create-react-app generated project.

**Comment by vkarpov15** [comment]
:man_facepalming: what version of create-react-app, what commands did you use to scaffold the project, what does "pretty much the same" mean, and what's the command you use to run your tests? Please provide exact answers, vague intimations aren't helpful.

**Comment by abubakkarbutt** [comment]
create-react-app --version
1.4.3

To run tests
`yarn test`

**Comment by abubakkarbutt** [comment]
`create-react-app app-name`
I used this command to generate my application project.

**Comment by abubakkarbutt** [comment]
I didn't changed configurations related to tests. I just changed the dev and build configurations which have no effect on running tests. So my test configurations are same as in newly generated project.

**Comment by lineus** [comment]
Here is how I duplicated this using create-react-app@1.4.3 and mongoose@5.0.6:
```
npm install create-react-app@1.4.3 
../node_modules/.bin/create-react-app testing
mv ../6141.js ./testing/src/
vi ./testing/src/6141.js // added a fn wrapper around my original code above and added export default 
vi ./testing/src/App.js // imported the fn from 6141.js and added a call to fn()
npm run test  // selected all from the test menu
 PASS  src/App.test.js
  ✓ renders without crashing (34ms)

Test Suites: 1 passed, 1 total
Tests:       1 passed, 1 total
Snapshots:   0 total
Time:        1.573s, estimated 2s
Ran all test suites.

  console.log src/6141.js:18
    version:  5.0.6

  console.log src/6141.js:19
    { name: 'John Doe',
      pets: [ {}, {}, {} ],
      _id: 5a8e9fa3bcbb0e79bb3d5a9f }
```
it's worth noting that before I got to the test step I tried to build the project:
```
npm run build // after the steps involving 6141.js in the last code block in this comment.
InspiredMacPro:testing lineus$ npm run build

> testing@0.1.0 build /Users/lineus/dev/Help/5/6141/testing
> react-scripts build

Creating an optimized production build...
Failed to compile.

Failed to minify the code from this file: 

 	/Users/lineus/dev/Help/5/node_modules/mongoose/lib/utils.js:15 

Read more here: http://bit.ly/2tRViJ9

npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! testing@0.1.0 build: `react-scripts build`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the testing@0.1.0 build script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/lineus/.npm/_logs/2018-02-22T10_53_44_670Z-debug.log
```
nothing in the the debug.log is jumping out at me yet but I'll keep digging into it to see if it's relevant. I'll also include the debug log here in case you see something I don't:
```
InspiredMacPro:testing lineus$ cat /Users/lineus/.npm/_logs/2018-02-22T10_53_44_670Z-debug.log
0 info it worked if it ends with ok
1 verbose cli [ '/Users/lineus/.nvm/versions/node/v9.2.0/bin/node',
1 verbose cli   '/Users/lineus/.nvm/versions/node/v9.2.0/bin/npm',
1 verbose cli   'run',
1 verbose cli   'build' ]
2 info using npm@5.5.1
3 info using node@v9.2.0
4 verbose run-script [ 'prebuild', 'build', 'postbuild' ]
5 info lifecycle testing@0.1.0~prebuild: testing@0.1.0
6 info lifecycle testing@0.1.0~build: testing@0.1.0
7 verbose lifecycle testing@0.1.0~build: unsafe-perm in lifecycle true
8 verbose lifecycle testing@0.1.0~build: PATH: /Users/lineus/.nvm/versions/node/v9.2.0/lib/node_modules/npm/bin/node-gyp-bin:/Users/lineus/dev/Help/5/6141/testing/node_modules/.bin:/Users/lineus/.nvm/versions/node/v9.2.0/bin:/Users/lineus/bin:/opt/local/bin:/bin:/sbin:/usr/bin:/usr/sbin:/usr/local/bin:/usr/local/sbin:/opt/local/sbin
9 verbose lifecycle testing@0.1.0~build: CWD: /Users/lineus/dev/Help/5/6141/testing
10 silly lifecycle testing@0.1.0~build: Args: [ '-c', 'react-scripts build' ]
11 silly lifecycle testing@0.1.0~build: Returned: code: 1  signal: null
12 info lifecycle testing@0.1.0~build: Failed to exec build script
13 verbose stack Error: testing@0.1.0 build: `react-scripts build`
13 verbose stack Exit status 1
13 verbose stack     at EventEmitter.<anonymous> (/Users/lineus/.nvm/versions/node/v9.2.0/lib/node_modules/npm/node_modules/npm-lifecycle/index.js:280:16)
13 verbose stack     at EventEmitter.emit (events.js:159:13)
13 verbose stack     at ChildProcess.<anonymous> (/Users/lineus/.nvm/versions/node/v9.2.0/lib/node_modules/npm/node_modules/npm-lifecycle/lib/spawn.js:55:14)
13 verbose stack     at ChildProcess.emit (events.js:159:13)
13 verbose stack     at maybeClose (internal/child_process.js:943:16)
13 verbose stack     at Process.ChildProcess._handle.onexit (internal/child_process.js:220:5)
14 verbose pkgid testing@0.1.0
15 verbose cwd /Users/lineus/dev/Help/5/6141/testing
16 verbose Darwin 15.6.0
17 verbose argv "/Users/lineus/.nvm/versions/node/v9.2.0/bin/node" "/Users/lineus/.nvm/versions/node/v9.2.0/bin/npm" "run" "build"
18 verbose node v9.2.0
19 verbose npm  v5.5.1
20 error code ELIFECYCLE
21 error errno 1
22 error testing@0.1.0 build: `react-scripts build`
22 error Exit status 1
23 error Failed at the testing@0.1.0 build script.
23 error This is probably not a problem with npm. There is likely additional logging output above.
24 verbose exit [ 1, true ]
InspiredMacPro:testing lineus$ 
```

**Comment by lineus** [comment]
I rolled back to mongoose@5.0.4 and re-ran the test, it's output:
```
 PASS  src/App.test.js
  ✓ renders without crashing (31ms)

Test Suites: 1 passed, 1 total
Tests:       1 passed, 1 total
Snapshots:   0 total
Time:        1.794s
Ran all test suites.

  console.log src/6141.js:18
    version:  5.0.4

  console.log src/6141.js:19
    { pets: 
       [ { _id: 5a8ea3df3d265a7f4aaf0c09, name: 'Parrot' },
         { _id: 5a8ea3df3d265a7f4aaf0c08, name: 'Mouse' },
         { _id: 5a8ea3df3d265a7f4aaf0c07, name: 'Dog' } ],
      _id: 5a8ea3df3d265a7f4aaf0c06,
      name: 'John Doe' }
```

**Comment by lineus** [comment]
it's worth noting that before I got to the test step I tried to build the project:
```
npm run build // after the steps involving 6141.js in the last code block in this comment.
InspiredMacPro:testing lineus$ npm run build

> testing@0.1.0 build /Users/lineus/dev/Help/5/6141/testing
> react-scripts build

Creating an optimized production build...
Failed to compile.

Failed to minify the code from this file: 

 	/Users/lineus/dev/Help/5/node_modules/mongoose/lib/utils.js:15 

Read more here: http://bit.ly/2tRViJ9

npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! testing@0.1.0 build: `react-scripts build`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the testing@0.1.0 build script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/lineus/.npm/_logs/2018-02-22T10_53_44_670Z-debug.log
```
nothing in the the debug.log is jumping out at me yet but I'll keep digging into it to see if it's relevant. I'll also include the debug log here in case you see something I don't:
```
InspiredMacPro:testing lineus$ cat /Users/lineus/.npm/_logs/2018-02-22T10_53_44_670Z-debug.log
0 info it worked if it ends with ok
1 verbose cli [ '/Users/lineus/.nvm/versions/node/v9.2.0/bin/node',
1 verbose cli   '/Users/lineus/.nvm/versions/node/v9.2.0/bin/npm',
1 verbose cli   'run',
1 verbose cli   'build' ]
2 info using npm@5.5.1
3 info using node@v9.2.0
4 verbose run-script [ 'prebuild', 'build', 'postbuild' ]
5 info lifecycle testing@0.1.0~prebuild: testing@0.1.0
6 info lifecycle testing@0.1.0~build: testing@0.1.0
7 verbose lifecycle testing@0.1.0~build: unsafe-perm in lifecycle true
8 verbose lifecycle testing@0.1.0~build: PATH: /Users/lineus/.nvm/versions/node/v9.2.0/lib/node_modules/npm/bin/node-gyp-bin:/Users/lineus/dev/Help/5/6141/testing/node_modules/.bin:/Users/lineus/.nvm/versions/node/v9.2.0/bin:/Users/lineus/bin:/opt/local/bin:/bin:/sbin:/usr/bin:/usr/sbin:/usr/local/bin:/usr/local/sbin:/opt/local/sbin
9 verbose lifecycle testing@0.1.0~build: CWD: /Users/lineus/dev/Help/5/6141/testing
10 silly lifecycle testing@0.1.0~build: Args: [ '-c', 'react-scripts build' ]
11 silly lifecycle testing@0.1.0~build: Returned: code: 1  signal: null
12 info lifecycle testing@0.1.0~build: Failed to exec build script
13 verbose stack Error: testing@0.1.0 build: `react-scripts build`
13 verbose stack Exit status 1
13 verbose stack     at EventEmitter.<anonymous> (/Users/lineus/.nvm/versions/node/v9.2.0/lib/node_modules/npm/node_modules/npm-lifecycle/index.js:280:16)
13 verbose stack     at EventEmitter.emit (events.js:159:13)
13 verbose stack     at ChildProcess.<anonymous> (/Users/lineus/.nvm/versions/node/v9.2.0/lib/node_modules/npm/node_modules/npm-lifecycle/lib/spawn.js:55:14)
13 verbose stack     at ChildProcess.emit (events.js:159:13)
13 verbose stack     at maybeClose (internal/child_process.js:943:16)
13 verbose stack     at Process.ChildProcess._handle.onexit (internal/child_process.js:220:5)
14 verbose pkgid testing@0.1.0
15 verbose cwd /Users/lineus/dev/Help/5/6141/testing
16 verbose Darwin 15.6.0
17 verbose argv "/Users/lineus/.nvm/versions/node/v9.2.0/bin/node" "/Users/lineus/.nvm/versions/node/v9.2.0/bin/npm" "run" "build"
18 verbose node v9.2.0
19 verbose npm  v5.5.1
20 error code ELIFECYCLE
21 error errno 1
22 error testing@0.1.0 build: `react-scripts build`
22 error Exit status 1
23 error Failed at the testing@0.1.0 build script.
23 error This is probably not a problem with npm. There is likely additional logging output above.
24 verbose exit [ 1, true ]
InspiredMacPro:testing lineus$ 
```

**Comment by vkarpov15** [comment]
I have no clue as to why this started in 5.0.5, but the root cause of this issue is the [webpack document provider](https://github.com/Automattic/mongoose/commit/eda8c551880352ac713e252fcbc0c2fe09e9cbc3), because webpack with default configs will always pick the `.web.js` file over the `.js` file:

```
$ cat test.web.js 
console.log('hello');
$ cat test.js
console.log('hi');
$ cat index.js 
import test from './test';
$ ./node_modules/.bin/webpack --entry ./index.js --resolve-extensions ".web.js,.js" -o ./out.js
Hash: 2c3cb3ae7cf56fad4865
Version: webpack 4.0.0
Time: 197ms
Built at: 2018-2-26 11:56:50
 Asset       Size  Chunks             Chunk Names
out.js  607 bytes       0  [emitted]  null
Entrypoint null = out.js
   [0] ./test.web.js 22 bytes {0} [built]
   [1] ./index.js 27 bytes {0} [built]

WARNING in configuration
The 'mode' option has not been set. Set 'mode' option to 'development' or 'production' to enable defaults for this environment. 
$ cat out.js 
!function(e){var n={};function t(r){if(n[r])return n[r].exports;var o=n[r]={i:r,l:!1,exports:{}};return e[r].call(o.exports,o,o.exports,t),o.l=!0,o.exports}t.m=e,t.c=n,t.d=function(e,n,r){t.o(e,n)||Object.defineProperty(e,n,{configurable:!1,enumerable:!0,get:r})},t.r=function(e){Object.defineProperty(e,"__esModule",{value:!0})},t.n=function(e){var n=e&&e.__esModule?function(){return e.default}:function(){return e};return t.d(n,"a",n),n},t.o=function(e,n){return Object.prototype.hasOwnProperty.call(e,n)},t.p="",t(t.s=1)}([function(e,n){console.log("hello")},function(e,n,t){"use strict";t.r(n);t(0)}]);$ 
$ 
```

This means that with the default create-react-app webpack config, mongoose will run partially in browser mode and partially in node mode.

TLDR, we'll get rid of the .web.js file. To use mongoose in browser mode, you should do `const mongoose = require('mongoose/lb/browser');` . We're going to add a shorthand so you can do `const mongoose = require('mongoose/browser');` in 5.0.8.
